cmd/gangue/gangue.go

```diff
@@ -0,0 +1,139 @@
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+
+	"github.com/spf13/cobra"
+
+	"github.com/coreos/mantle/auth"
+	"github.com/coreos/mantle/cli"
+	"github.com/coreos/mantle/sdk"
+)
+
+var (
+	root = &cobra.Command{
+		Use:   "gangue",
+		Short: "Google Storage download and verification tool",
+	}
+
+	get = &cobra.Command{
+		Use:   "get [url] [path]",
+		Short: "download and verify a file from Google Storage",
+		Run:   run,
+	}
+
+	gpgKeyFile, jsonKeyFile, output, source string
+	serviceAuth, verify                     bool
+)
+
+func validateGSURL(rawURL string) error {
+	parsedURL, err := url.Parse(rawURL)
+	if err != nil {
+		return err
+	}
+	if parsedURL.Scheme != "gs" {
+		return fmt.Errorf("URL missing gs:// scheme: %v", rawURL)
+	}
+	if parsedURL.Host == "" {
+		return fmt.Errorf("URL missing bucket name %v", rawURL)
+	}
+	if parsedURL.Path == "" {
+		return fmt.Errorf("URL missing file path %v", rawURL)
+	}
+	if parsedURL.Path[len(parsedURL.Path)-1] == '/' {
+		return fmt.Errorf("URL must not be a directory path %v", rawURL)
+	}
+	return nil
+}
+
+func run(cmd *cobra.Command, args []string) {
+	var client *http.Client
+
+	if len(args) == 2 {
+		source = args[0]
+		output = args[1]
+	} else if len(args) == 1 {
+		source = args[0]
+		output, _ = os.Getwd()
+	} else {
+		fmt.Fprintf(os.Stderr, "Expected one or two arguments\n")
+		os.Exit(1)
+	}
+
+	// Perform some basic sanity checks on the options
+	if output == "" {
+		fmt.Fprintf(os.Stderr, "The output file path is required\n")
+		os.Exit(1)
+	}
+	err := validateGSURL(source)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%v\n", err)
+		os.Exit(1)
+	}
+
+	// If the output path exists and is a directory, keep the file name
+	if stat, err := os.Stat(output); err == nil && stat.IsDir() {
+		output = path.Join(output, path.Base(source))
+	}
+
+	// Authenticate with Google
+	if serviceAuth {
+		client = auth.GoogleServiceClient()
+	} else if jsonKeyFile != "" {
+		b, err := ioutil.ReadFile(jsonKeyFile)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "%v\n", err)
+			os.Exit(1)
+		}
+		client, err = auth.GoogleClientFromJSONKey(b)
+	} else {
+		client, err = auth.GoogleClient()
+	}
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%v\n", err)
+		os.Exit(1)
+	}
+
+	// Download the file and verify it (unless disabled)
+	if verify {
+		err = sdk.UpdateSignedFile(output, source, client, gpgKeyFile)
+	} else {
+		err = sdk.UpdateFile(output, source, client)
+	}
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%v\n", err)
+		os.Exit(1)
+	}
+}
+
+func main() {
+	bv := get.PersistentFlags().BoolVar
+	sv := get.PersistentFlags().StringVar
+
+	bv(&serviceAuth, "service-auth", false, "use non-interactive auth when running within GCE")
+	sv(&jsonKeyFile, "json-key", "", "use a service account's JSON key for authentication")
+	bv(&verify, "verify", true, "use GPG verification")
+	sv(&gpgKeyFile, "verify-key", "", "PGP public key file to verify signatures")
```

Comment:
Perhaps explicitly state that this is an alternative key, since our default image signing key will be used if blank.